[camera_android_camerax] Add `AGENTS.md` (#12066)
flutter/packages: packages/camera/camera_android_camerax/AGENTS.md

Reviewer: High level feedback this is much longer than an initial draft I thinks should be. Following the guidance in https://www.kaggle.com/whitepaper-the-new-SDLC-with-vibe-coding I think we should try to make this as short as we can since every token is an overhead in every conversation.

Author: Might have gone extreme in the other direction but the agent I edited with made it seem like I can get rid of all the wordiness and focus on the core operations/skills it needs to know so edited it down to this. PTAL!

# Agent Guide for camera_android_camerax

This document provides context, architectural details, and core workflows
for making high quality contributions to the `camera_android_camerax` package.

## Architectural Overview

The `camera_android_camerax` package is the Android platform implementation

> of the `camera` plugin, using the **Android Jetpack CameraX** library.

Reviewer: ```suggestion
of the (camera)[../camera] plugin, using the (**Android Jetpack CameraX**)[https://developer.android.com/jetpack/androidx/releases/camera] library.
```

### ProxyApi Binding System

> This plugin utilizes **Pigeon's ProxyApi** to bind Dart objects directly

Reviewer: ```suggestion
This plugin utilizes (**Pigeon's ProxyApi**)[../../pigeon] to bind Dart objects directly
```

to native Android CameraX Java/Kotlin objects.

- **Pigeon Definition**: The API surfaces are defined in
  [pigeons/camerax_library.dart](pigeons/camerax_library.dart).
- **Generated Code**:
  - Dart proxy classes:
    [lib/src/camerax_library.g.dart](lib/src/camerax_library.g.dart)
  - Kotlin proxy classes:
    [android/src/main/java/io/flutter/plugins/camerax/CameraXLibrary.g.kt](android/src/main/java/io/flutter/plugins/camerax/CameraXLibrary.g.kt)
- **Plugin Entry Point**:
  [lib/src/android_camera_camerax.dart](lib/src/android_camera_camerax.dart)
  implements `CameraPlatform` using the generated proxy classes.

## Required Steps Before Making Any Changes

> Before making any changes, run the

Reviewer: ```suggestion
Before working, run the
```

> [check-readiness skill](.agents/skills/check-readiness/SKILL.md)

Author: I don't think this is right.

> [check-readiness skill](.agents/skills/check-readiness/SKILL.md)

Reviewer: Confirmed you are correct and gemini is incorrect https://github.com/flutter/packages/tree/main/packages/camera/camera_android_camerax/.agents/skills/check-readiness

> to ensure your environment has all the required tooling.

Reviewer: ```suggestion
to ensure your environment has all the expected tooling.
```

Depending on the task the tooling might not be required.

> ## Required Steps After Making Any Changes

Reviewer: I think this section should not be in agents.md but instead in a skill. But if it is helpful now then we can keep it.

     - Dart unit tests.
     - Android native unit tests.
     - Integration tests.
3. **Static Analysis**:
   - Run the
     [dart-run-static-analysis skill](.agents/skills/dart-run-static-analysis/SKILL.md)
     to analyze the code and apply automated fixes.

## Code Generation

When you modify the Pigeon API
[pigeons/camerax_library.dart](pigeons/camerax_library.dart),
you must regenerate the respective code to update the Dart proxy class
[lib/src/camerax_library.g.dart](lib/src/camerax_library.g.dart)
and the Kotlin proxy class
[android/src/main/java/io/flutter/plugins/camerax/CameraXLibrary.g.kt](android/src/main/java/io/flutter/plugins/camerax/CameraXLibrary.g.kt).
Run from this directory:

```bash
dart run pigeon --input pigeons/camerax_library.dart
```

Because tests in this package use mockito for mocking, you must also
regenerate the Mockito mocks that are used for unit testing any changes
you make by running in this directory:

```bash
dart run build_runner build -d
```

For more in-depth guidance on creating or updating mocks, refer to the
[dart-generate-test-mocks skill](.agents/skills/dart-generate-test-mocks/SKILL.md).

## Running Tests

> When you make a change, add a test if you can (either a Dart unit test,

Reviewer: *gut feeling*, I think this should say 
```suggestion
When you make a change, add a test if other similar code has tests(either a Dart unit test,
```

Android native unit test, or a Flutter integration test). Regardless of
whether you add a test or not, all tests must pass after you make changes.

> How to run the tests:

Reviewer: "How to run the tests" feels like a skill more importantly it feels like agent instructions that will drift from the truth without something to inform them. 

Where do you think this knowledge should live long term? Could we get away with describing the types of tests and not say how to run them?

### Dart Unit Tests

Dart unit tests are located in [test/](./test/).
To run them from this directory:

```bash
dart run ../../../script/tool/bin/flutter_plugin_tools.dart dart-test --packages=camera_android_camerax
```

### Android Native Unit Tests

Android unit tests are located in [android/src/test/](./android/src/test/).
To run them from this directory:

```bash
dart run ../../../script/tool/bin/flutter_plugin_tools.dart native-test --android --packages=camera_android_camerax
```

### Flutter Integration Tests

Integration tests are located in
[example/integration_test/integration_test.dart](./example/integration_test/integration_test.dart).
For guidance on writing new integration tests or using MCP to explore
UI flows interactively, refer to the
[flutter-add-integration-test skill](.agents/skills/flutter-add-integration-test/SKILL.md).
They require an Android emulator or physical device to be connected.
To check if this is true, run `flutter devices` and verify that an

> Android device/emulator is listed. If not, prompt the user to start one.

Reviewer: running flutter devices does not check if a physical device or emulator is required.

If so, run from this directory:

```bash
dart run ../../../script/tool/bin/flutter_plugin_tools.dart integration-test --android --packages=camera_android_camerax
```

> ## Required Steps Before Pushing

Reviewer: I think we might want to change the tone of this section. We should introduce the skill, let the skill describe what it does, and describe when it makes sense to run that skill and why it is important to do so. 

For example the skill gives feedback on changes that are required before landing. By not running the skill it is likely that either presubmits will fail or that code review feedback will block the change which is why it is important to run before pushing.
